R/desc_utils.R

```diff
@@ -0,0 +1,208 @@
+#'
+#' d <- desc::desc(
+#'   file = verdepcheck:::local_description(
+#'     list(
+#'       dplyr = "Import",
+#'       tibble = "Import",
+#'       pkgdepends = "Import"
+#'     ),
+#'     remotes = c(
+#'       "tidyverse/dplyr@*release",
+#'       "tidyverse/tibble@*release",
+#'       "r-lib/pkgdepends@*release"
+#'     ),
+#'     need_verdepcheck = c(
+#'       "dplyr",
+#'       "tibble=tidyverse/tibble@v3.2.1"
+#'     )
+#'   )
+#' )
+#' verdepcheck:::desc_remotes_cleanup(d)
+desc_remotes_cleanup <- function(d) {
+  if (length(get_desc_field_pkgs(d)) == 0) {
+    return(d)
+  }
+  # Parse the `Config/Needs/verdepcheck` to retrieve references and extract package names
+  desc_field_refs <- pkgdepends::parse_pkg_refs(get_desc_field_pkgs(d))
+  desc_field_names <- map_key_character(desc_field_refs, "package")
+
+  # Parse the remotes to retrieve the package names
+  remotes_refs <- pkgdepends::parse_pkg_refs(d$get_remotes())
+  remotes_names <- map_key_character(remotes_refs, "package")
+
+  # Add to remotes `Config/Needs/verdepcheck` that resolve to a remote_ref_github
+  desc_field_include_ix <- vapply(desc_field_refs, inherits, logical(1), "remote_ref_github") &
+    desc_field_names %in% remotes_names
+
+  # Only keep previous remotes that are not defined in `Config/Needs/verdepcheck`
+  remotes_include_ix <- remotes_names %in% setdiff(remotes_names, desc_field_names)
+
+  # Create new list of references that will be used as "Remotes"
+  new_remotes <- c(
+    map_key_character(desc_field_refs[desc_field_include_ix], "ref"),
+    map_key_character(remotes_refs[remotes_include_ix], "ref")
+  )
+
+  new_d <- d$clone()
+  # Remove all remotes and override it
+  new_d$clear_remotes()
+
+  # Return clause without Remotes section if none should be kept
+  if (is.null(new_remotes) || length(new_remotes) == 0) {
+    return(new_d)
+  }
+  new_d$set_remotes(new_remotes)
+  new_d
+}
+
+#' Set `"Config/Needs/verdepcheck"` section into the `desc` object if not empty else clear this section.
+#' @keywords internal
+desc_cond_set_refs <- function(d, refs) {
+  if (length(refs)) {
+    d$set_list(.desc_field, refs)
+  } else {
+    d$del(.desc_field)
+  }
+  return(invisible(d))
+}
+
+#' Create `installation_plan` object from `desc` object
+#' @importFrom pkgdepends new_pkg_deps new_pkg_installation_proposal
+#' @keywords internal
+desc_to_ip <- function(d, config) {
+  temp_desc <- tempfile()
+  d$write(temp_desc)
+
+  pkgdepends::new_pkg_installation_proposal(
+    refs = paste0("deps::", temp_desc),
+    config = config
+  )
+}
+
+#' Get package version from description
+#'
+#' @keywords internal
+#'
+#' @examples
+#' d <- desc::desc(cmd = "!new")
+#'
+#' d$set_dep("magrittr", type = "Imports", version = "*")
+#' verdepcheck:::version_from_desc("magrittr", d)
+#'
+#' d$set_dep("magrittr", type = "Imports", version = ">= 1.5")
+#' verdepcheck:::version_from_desc("magrittr", d)
+version_from_desc <- function(pkg_name, desc) {
```

Comment:
[low priority] - reorder arguments to make `desc` first; also: all functions above use `d` instead of `desc` - both names are good but we need to stay consistent

Reply:
done in 95f7989